crates/autopilot/src/domain/competition/participation_guard/statistics.rs

```diff
@@ -0,0 +1,511 @@
+use {
+    crate::{
+        arguments::StatisticsBasedSolverParticipationGuardConfig,
+        domain::{Metrics, auction, competition, eth},
+        infra::{self, solvers::dto},
+    },
+    ethrpc::block_stream::CurrentBlockWatcher,
+    std::{
+        collections::{HashMap, HashSet, VecDeque},
+        sync::Arc,
+    },
+    tokio::sync::{Mutex, mpsc},
+};
+
+/// Tracks the solver settlement statistics and decides whether the solver
+/// should be banned based on the failure rate and consecutive failed
+/// settlements.
+#[derive(Clone)]
+pub(super) struct SolverValidator(Arc<Inner>);
+
+struct Inner {
+    persistence: infra::Persistence,
+    /// A map of banned solver addresses and the corresponding number of
+    /// settlements still need to skip before the solver can participate in
+    /// the competition again.
+    banned_solvers: dashmap::DashMap<eth::Address, u32>,
+    /// Stop banning solvers when the number of remaining active
+    /// solvers hits this threshold. This is crucial to prevent the protocol
+    /// from blocking itself by banning all the solvers since the recover
+    /// mechanism requires new competitions.
+    min_active_solvers_threshold: u32,
+    config: StatisticsBasedSolverParticipationGuardConfig,
+    drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+    competitions_tracker: Mutex<CompetitionsTracker>,
+}
+
+/// Keeps track of the solver's competition submission statistics.
+#[derive(Clone, Debug, Default)]
+struct SolverCompetitionStats {
+    total: u32,
+    failed: u32,
+}
+
+impl SolverCompetitionStats {
+    fn increment(&mut self, success: bool, high_failure_threshold: f64) {
+        if success {
+            self.total += 1;
+            // Recover fast in case the solver is banned.
+            if self.failed > 0 {
+                // Keep the failure rate at the threshold level, so the solver is
+                // able to participate again. And at the same time, on the next failure it gets
+                // banned again.
+                while self.failure_rate() >= high_failure_threshold {
+                    self.total += 1;
+                }
```

Comment:
This seems super sketchy. The higher the `total` the less individual successes or failures change the ratio which seems problematic.
If the idea is to implement a sliding window I would expect `total` to be constant so `failed` would have to be reduced instead. Although I'm not sure this is even overall the best solution. Have to think more about that.